sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ExpandShufflePartitions.scala

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.adaptive
+
+import org.apache.spark.sql.catalyst.plans.physical.SinglePartition
+import org.apache.spark.sql.execution.SparkPlan
+import org.apache.spark.sql.execution.exchange.{EnsureRequirements, REPARTITION_BY_COL, REPARTITION_BY_NONE, ShuffleExchangeLike, ShuffleOrigin}
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * A rule to expand the shuffle partitions based on the map output statistics, which can
+ * avoid data skew that hurt performance.
+ *
+ * We use ADVISORY_PARTITION_SIZE_IN_BYTES size to decide if a partition should be expanded.
+ * Let's say we have 3 maps with 3 shuffle partitions, and assuming r1 has data skew issue.
+ * the map side looks like:
+ *   m0:[b0, b1, b2], m1:[b0, b1, b2], m2:[b0, b1, b2]
+ * and the reduce side looks like:
+ *                          (without this rule) r1[m0-b1, m1-b1, m2-b1]
+ *                              /                              \
+ *   r0:[m0-b0, m1-b0, m2-b0], r1:[m0-b1], r2:[m1-b1], r3:[m2-b1], r4[m0-b2, m1-b2, m2-b2]
+ */
+object ExpandShufflePartitions extends CustomShuffleReaderRule {
+  override def supportedShuffleOrigins: Seq[ShuffleOrigin] =
+    Seq(REPARTITION_BY_COL, REPARTITION_BY_NONE)
```

Comment:
why can we accept `REPARTITION_BY_COL`? If people do `re.repartition($"a")`, we should make sure the output is hash partitioned by column a, isn't it?

Reply:
I think it only makes sense if the `repartition($"a")` is added by the framework to optimize table insertion, not added by users.

Reply:
This makes me think that if we need a new operator to do the repartition for partitioned table insertion (non partitioned table can use the existing operator, thanks to https://github.com/apache/spark/commit/ce1636948b1fb8cfb8cc921896dc003949da1085), and assign it a new shuffle origin.

Reply:
cc @wangyum

Reply:
> If people do df.repartition($"a"), we should make sure the output is hash partitioned by column a, isn't it

yea, we should promise this.

> I think it only makes sense if the repartition($"a") is added by the framework to optimize table insertion, not added by users.

yea, we can use a new operator and shuffle origin to distinguish if it is added by user or framework. Then only optimize the operator added by framework.

The origin idea of this PR is that add a config to let user decide if `repartition($"a")` can expand partitions which break the semantics, so user can use it in SQL query easily. After some thought maybe it's better to add a new hint that can support expand partitions ?

Reply:
We can start with the new operator first, and think of the user-facing API later. Maybe we don't need a user-facing API and the new operator can only be used by the table insertion optimizer rule.

Reply:
@cloud-fan  created [#32932](https://github.com/apache/spark/pull/32932). in order to show the usage, that PR added a new hint.

Reply:
Do we try to improve this case?
```sql
CREATE TABLE t1 USING parquet PARTITIONED BY (p2)
AS
(SELECT id, id % 1000 AS p1, if(id < 3000000, 1, id % 100) AS p2 FROM range(3100000) distribute by p2)
```

Reply:
yes, here are two approach to optimize this case you give:
* this PR can optimize it directly using a new config, but a potential issue is the `distribute by` semantics of output partitioning will be changed.
*  [#32932](https://github.com/apache/spark/pull/32932) is going to use a new operator which does not guarantee the output partitioning, then we can optimize the new operator safely.